core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -518,8 +525,87 @@ primitiveTypeAssociations = [
     (''Prims.ByteString, pure $ appSize haskellBSSize $ PLC.TyBuiltin () PLC.TyByteString)
     ]
 
+-- | The function 'error :: forall a . () -> a'.
 errorFunc :: Quote (PLC.Term PLC.TyName PLC.Name ())
-errorFunc = freshTyName () "e" >>= \n -> pure $ PLC.TyAbs () n (PLC.Type ()) $ PLC.Error () (PLC.TyVar () n)
+errorFunc = do
+    n <- freshTyName () "e"
+    -- see Note [Value restriction]
+    mangleTyAbs $ PLC.TyAbs () n (PLC.Type ()) (PLC.Error () (PLC.TyVar () n))
+
+-- Value restriction
+
+{- Note [Value restriction]
+Plutus Core has the traditional *value restriction* on type abstractions - namely, the
+body of a type abstraction must be a value.
+
+This causes problems for us because *Haskell* has no such thing.
+
+There are two approaches to solving this problem. Currently we use "passing on the value restriction",
+but I include a description of mangling to illustrate why the current solution is preferable.
+
+-- Mangling
+
+We can get around this by delaying the body of all our abstractions, thus making them lambdas, which are values.
+This then means that we need to change the *types* of all foralls to include the unit argument,
+and all instantiations to force the value.
+
+We need to do this everywhere so that the translation of the users' program remains
+well-typed. Consider
+
+runST :: (forall s. ST s a) -> a
```

Comment:
I'm not sure this is a good example, because we compile Core and in Core `ST` is functional.
Though, still illustrates the problem.